javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -1387,23 +1265,158 @@ public String toString() {
         return mkString(stringPrefix() + "(", ", ", ")");
     }
 
-    private static <T> Vector<T> wrap(HashArrayMappedTrie<Integer, T> trie) {
-        return trie.isEmpty() ? empty() : new Vector<>(trie);
-    }
-}
+    static final class VectorTree<T> implements Serializable {
+        private static final long serialVersionUID = 1L;
 
-interface VectorModule {
+        private static final VectorTree<?> EMPTY = new VectorTree<>(emptyArray(), 0, 0, 0);
+        @SuppressWarnings("unchecked")
+        static <T> VectorTree<T> emptyTree() { return (VectorTree<T>) EMPTY; }
 
-    final class Combinations {
+        private final Object[] array;
+        private final int offset, length;
+        private final int depthShift;
 
-        static <T> Vector<Vector<T>> apply(Vector<T> elements, int k) {
-            if (k == 0) {
-                return Vector.of(Vector.empty());
+        private VectorTree(Object[] array, int offset, int length, int depthShift) {
+            this.array = array;
+            this.offset = offset;
+            this.length = length;
+            this.depthShift = depthShift;
+
+            assert length() <= treeSize(branchingFactor(), depthShift);
+        }
+
+        static int treeSize(int branchCount, int depthShift) {
```

Comment:
I would restrict all methods and fields to `private` access if possible. In most cases white-box unit tests are not really necessary. All branches can also be tested by using public API.